api/v1alpha3/conversion_test.go

```diff
@@ -77,13 +77,23 @@ func BootstrapFuzzFuncs(_ runtimeserializer.CodecFactory) []interface{} {
 	}
 }
 
+func MachineSetFuzzFuncs(_ runtimeserializer.CodecFactory) []interface{} {
+	return []interface{}{
+		MachineSetFuzzer,
+	}
+}
+
 func BootstrapFuzzer(obj *Bootstrap, c fuzz.Continue) {
 	c.FuzzNoCustom(obj)
 
 	// Bootstrap.Data has been removed in v1alpha4, so setting it to nil in order to avoid v1alpha3 --> v1alpha4 --> v1alpha3 round trip errors.
 	obj.Data = nil
 }
 
+func MachineSetFuzzer(obj *v1alpha4.MachineSet, c fuzz.Continue) {
+	obj.Status.DeletingReplicas = 0
+}
```

Comment:
This is a new field and should be preserved during conversion, rather than set to 0 here.

To preserve the field, in `conversion.go` we need to adjust the ConvertTo and ConvertFrom functions to look like the ones for `MachineDeployment` object.

- `ConvertFrom` needs:
```go
	// Preserve Hub data on down-conversion except for metadata
	if err := utilconversion.MarshalData(src, dst); err != nil {
		return err
	}
```
- `ConvertTo` needs:
```go
	// Manually restore data.
	restored := &v1alpha4.MachineDeployment{}
	if ok, err := utilconversion.UnmarshalData(src, restored); err != nil || !ok {
		return err
	}
```
\+ code to actually restore the field value.

Reply:
a big thank you for pointing this out for me! This was my first time working with the process of upgrading CRD versions and this was very helpful to me understanding more of that process (and spelling it out or else I would've been lost for a long time).

Looks like based off the other suggestion this might just all go away (if we just query machines when looking at old replicas)